src/visitors/semantic/completion/mod.rs

```diff
@@ -0,0 +1,152 @@
+use crate::visitors::semantic::utils;
+
+use flux::semantic::nodes::*;
+use flux::semantic::types::MonoType;
+use flux::semantic::walk::{Node, Visitor};
+
+#[derive(Default)]
+pub struct CompletableFinderState {
+    pub completables: Vec<Rc<dyn Completable>>,
+}
+
+pub struct CompletableFinderVisitor {
+    pub pos: Position,
+    pub state: Rc<RefCell<CompletableFinderState>>,
+}
+
+impl CompletableFinderVisitor {
+    pub fn new(pos: Position) -> Self {
+        CompletableFinderVisitor {
+            state: Rc::new(RefCell::new(
+                CompletableFinderState::default(),
+            )),
+            pos,
+        }
+    }
+}
+
+fn get_var_type(expr: Expression) -> Option<VarType> {
+    match expr {
+        Expression::Integer(_) => Some(VarType::Int),
+        Expression::Boolean(_) => Some(VarType::Bool),
+        Expression::Float(_) => Some(VarType::Float),
+        Expression::StringLit(_) => Some(VarType::String),
+        Expression::Array(_) => Some(VarType::Array),
+        Expression::Regexp(_) => Some(VarType::Regexp),
+        Expression::Duration(_) => Some(VarType::Duration),
+        Expression::Call(c) => {
+            let result_type = utils::follow_function_pipes(c);
+
+            match result_type {
+                MonoType::Int => Some(VarType::Int),
+                MonoType::Float => Some(VarType::Float),
+                MonoType::Bool => Some(VarType::Bool),
+                MonoType::Arr(_) => Some(VarType::Array),
+                MonoType::Duration => Some(VarType::Duration),
+                MonoType::Row(_) => Some(VarType::Row),
+                MonoType::String => Some(VarType::String),
+                MonoType::Uint => Some(VarType::Uint),
+                MonoType::Time => Some(VarType::Time),
+                _ => None,
+            }
+        }
+        _ => None,
+    }
+}
+
+impl<'a> Visitor<'a> for CompletableFinderVisitor {
+    fn visit(&mut self, node: Rc<Node<'a>>) -> bool {
+        let mut state = self.state.borrow_mut();
+        let loc = node.loc();
+        let pos = self.pos.clone();
+
+        if loc.start.line > pos.line + 1
+            || (loc.start.line == pos.line + 1
+                && loc.start.column > pos.character + 1)
+        {
+            return true;
+        }
+
+        if let Node::VariableAssgn(assgn) = node.as_ref() {
+            let id = assgn.id.clone();
+
+            if let Some(var_type) = get_var_type(assgn.init.clone()) {
+                (*state).completables.push(Rc::new(VarResult {
+                    var_type,
+                    name: id.name,
+                }))
+            }
+        }
+
+        true
+    }
+}
+
+#[derive(Clone)]
+enum VarType {
+    Int,
+    String,
+    Array,
+    Float,
+    Bool,
+    Duration,
+    Regexp,
+    Row,
+    Uint,
+    Time,
+}
+
+#[derive(Clone)]
+struct VarResult {
+    pub name: String,
+    pub var_type: VarType,
+}
+
+impl VarResult {
+    pub fn detail(&self) -> String {
```

Comment:
It's cleaner to implement a Display trait for the VarType.

Reply:
this is directly related to a field we return to the LSP client named detail, so I prefer to be explicit here.